The branch runs the nextflow pipeline with GATK bash scripts (instead of Sentieon). This is based off Dave's `newer_workflow_Dave` branch. Bash code comes from the WDL repo


# Haplotype Variant Calling with GATK4 on Nextflow
This repo is a variant calling pipeline that uses GATK4 in [Nextflow](https://www.nextflow.io/) workflow management language, for use in detecting single nucleotide polymorphisms (SNP) and short insertion deletion variants. It is a Nextflow-based implementation of [Cromwell-WDL based MayomicsVC workflow](https://github.com/ncsa/MayomicsVC).


# Synthetic data generation

The functionality of this pipeline has been tested using synthetic dataset generated by [neat-gen](https://github.com/zstephens/neat-genreads). A whole Exome Sequencing dataset of 30X of 5 batches (i.e. 5 lanes) was simulated via:

```
wget https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/truseq/truseq-exome-targeted-regions-manifest-v1-2-bed.zip
reference="/home/a-m/azzaea/varCall/genome/ucsc.hg19.fasta"
NumberOfChunks=5
NeatParams="-R 101 -c 100 -p 2 -t truseq-exome-targeted-regions-manifest-v1-2.bed --pe 300 30 --rng 100 --vcf"
for (( jobnumber=1; jobnumber<=${NumberOfChunks}; jobnumber++ )); do
        python -u ${NEAT2_path}/genReads.py -r ${reference} -o ${OutputChunksFolder}/${DatasetName} ${NeatParams} --job ${jobnumber} ${NumberOfChunks} >  ${OutputLogsFolder}/${DatasetName}.GenerateChunks.job_${jobnumber}_of_${NumberOfChunks}.log 
done
```


